Multiple Patient Address Initial Support

adunsulag opened the pull request:
Fixes #5367 

Added a new data type for LBF to support Address List data type.

Addresses can be added with multiple addresses supported.  Initial
framework for saving addresses is here as well s retrieving them.

Rolled out the initial contact and contact address classes that can
be used to support Patients.  Decided to leverage the ORM classes
for Contact and ContactAddress to quickly persist and model
the data.

This is a WIP draft for people to be able to see what I'm proposing.  It currently doesn't fetch the address information from the database but we're super close to being able to do that.  

I'll get back to this once we make it through the clinical quality measures stuff.

adunsulag commented:
@psoas This is what I've currently done.  I was nearly finished with it but things got busy with the Clinical Quality Measures stuff.  My hope is to wrap this up next week.  I wanted to push this out there so that you and others can have a chance to see it and what its doing.  It's probably a half a day to a day away from being finished and we'll be able to then bring it in.

bradymiller commented on library/layout.inc.php:
```diff
@@ -61,7 +63,8 @@
     "52" => xl("Previous Names"),
-    "53" => xl("Patient Encounters List")
+    "53" => xl("Patient Encounters List"),
+    "54" => xl("Address List")
```
![2DV-2](https://user-images.githubusercontent.com/278968/171539574-7a5f48b5-e5b2-4310-b634-22831f89dc78.gif)

bradymiller commented on library/templates/address_list_display.php:
```diff
@@ -0,0 +1,84 @@
+    <table class='table-text-addresses table table-sm'>
+        <thead class ="thead-light">
+        <tr>
+            <th width="400">Previous Addresses</th>
```
xlt

adunsulag replied:
Thanks brady, fixed this issue.

bradymiller commented on library/templates/address_list_display.php:
```diff
@@ -0,0 +1,84 @@
+
+        }
+        window.document.addEventListener('DOMContentLoaded', function() {
+            init("<?php echo $table_id; ?>", <?php echo json_encode($addresses); ?>);
```
would used js_escape() around both variables above (rather than json_encode since allows us to "fix" js escaping issues in a centralized function, if needed)

bradymiller replied:
bte, ok to keep json_encode for $addresses is this is meant for more than just escaping (ie. migrating php array to json)

adunsulag replied:
got the escaping in for table_id.

bradymiller commented on library/templates/address_list_form.php:
```diff
@@ -0,0 +1,278 @@
+                    <td width='50' colspan='1'><?php echo xlt("City"); ?></td>
+                    <td width='8' colspan='1'><?php echo xlt("State"); ?></td>
+                    <td width='12' colspan='1'><?php echo xlt("Postal Code"); ?></td>
+                    <td width='12' colspan='1'><?php echo xlt("Plus Four"); ?></td>
```
remove the Plus Four and instead store everything in Postal Code

adunsulag replied:
Made everything use Postal Code on the frontend.  I did put in logic to split the postal code into zip + plus_four if the country is USA when we are saving to the database.  Seems like that'd be a good thing for data searching, but not sure if that will break in other parts of the system.  If its an issue I can just store the data directly into postalcode.

bradymiller replied:
would just use postalcode since the plus_four is never used in OpenEMR at this time (at least I couldn't find a place it is used). then never need to have country specific logic.

bradymiller commented on library/templates/address_list_form.php:
```diff
@@ -0,0 +1,278 @@
+                <tr height='10'><td></td></tr>
+                <tr>
+                    <td width='50' colspan='1'><?php echo xlt("City"); ?></td>
+                    <td width='8' colspan='1'><?php echo xlt("State"); ?></td>
```
For City and Country would put them through the datatypes that are using in the current layouts for them (this will effectively be using a datatype within a datatype, which should work)

adunsulag replied:
Made this change but I couldn't call the datatype function due to some of the html so I used the generate lists functionality.  Downside is not having the 'Add' button to the datatype.  If that's critical I'll need to rework this. I also need to figure out how to keep the tab order @psoas was using originally now that we are generating the select dropdowns.

bradymiller replied:
@adunsulag and @psoas , Been thinking about this and do we really need to have City and Country come from a list in the more modern implementation of addresses (I do realize I am the one that recommended this above, but on further thought, this just seems really limiting when seems like a free text field makes more sense considering goal will be at some point to integrate with a address validator). thoughts?

bradymiller commented on library/templates/address_list_form.php:
```diff
@@ -0,0 +1,278 @@
+    function DeleteAddress(element){
+        var row_display_addresses = element.closest('tr.display_addresses');
+        var row_form_addresses = row_display_addresses.nextElementSibling;
+        let prompt = window.top.i18next.t("ARE YOU REALLY REALLY SURE?");
```
should be able to just make this:
```
let prompt = window.xl("ARE YOU REALLY REALLY SURE?");
```

adunsulag replied:
Fixed this.

bradymiller commented on sql/6_1_0-to-7_0_0_upgrade.sql:
```diff
@@ -675,3 +675,36 @@ INSERT INTO list_options ( list_id, option_id, title, seq, is_default ) VALUES (
+    `type` VARCHAR(255) NULL COMMENT '[Values: Home, Work, Physical, Mailing, Shipping]',
+    `notes` TEXT(255) NULL,
+    `status` BINARY(1) NULL COMMENT '[Values: Active, Inactive, etc]',
+    `is_primary` BINARY(1) NULL,
```
haven't seen use of BINARY for this before. The Status one especially where looks like there are those values going into binary.

adunsulag replied:
Had to change this to a char field because of ORDataObject funness.

bradymiller commented on sql/6_1_0-to-7_0_0_upgrade.sql:
```diff
@@ -675,3 +675,36 @@ INSERT INTO list_options ( list_id, option_id, title, seq, is_default ) VALUES (
+    `period_start` DATETIME NULL COMMENT 'Date the address became active',
+    `period_end` DATETIME NULL COMMENT 'Date the address became deactivated',
+    `inactivated_reason` VARCHAR(45) NULL DEFAULT NULL COMMENT '[Values: Moved, Mail Returned, etc]',
+    PRIMARY KEY (`contact_id`),
```
is this primary key supposed to be `id` ?

adunsulag replied:
Fixed this issue.

bradymiller commented on library/classes/Contact.class.php:
```diff
@@ -0,0 +1,95 @@
+<?php
```
would store new classes somewhere in src tree with name of file with class name only like Contact.php
(long term goal is all classes migrate to psr4 compatible in src tree including the legacy ones based on ORDataObject)

bradymiller commented:
@adunsulag , Just did a quick review. Nice stuff!! Will do another review when out of draft.

adunsulag commented:
Got the addresses saving and reloading now.  Still need to do a bunch more testing and adding in some of the items identified in #5367. I'm also keen on potentially bringing in @mdsupport's PR here #4862 so we can keep things idempotent with transactions.

Was hoping to finish this yesterday, but we have more pressing issues with the schematron fixes for CCDA, so I'll circle back around on this when I'm done with that.

sjpadgett commented on library/templates/address_list_form.php:
```diff
@@ -0,0 +1,322 @@
+$name_field_id = "form_" . $field_id_esc;
+$smallform = $smallform ?? '';
+?>
+<div>
+    <table class='form_addresses table table-sm' id="<?php echo attr($table_id); ?>">
+        <thead class ="thead-light">
+        <tr>
+            <th colspan="2">
+                <?php echo xlt("Additional Addresses"); ?>&nbsp;&nbsp;&nbsp;&nbsp;
+                <span class="fas fa-plus-square" style="color:#007bff;" onclick="AddAddress(event);return false"></span>
+            </th>
+        </tr>
+        </thead>
+    </table>
+
+</div>
```
I hate to be the ole fuddy-duddy but, our whole purpose of spending all the effort to refactor to bootstrap was for responsiveness. We've gradually been moving our LBF render engine to bootstrap with our eye set to refactor out all the table layouts into templates as plug in and reusable.

These forms are not too difficult to make into bootstrap forms with all the advantages that entails.

I'm not strongly opposed to the current form layouts however we are simply adding more work for later where this project could give us the opportunity to set a pattern for future development.

adunsulag replied:
I don't mind refactoring the forms, my primary motivation was first taking what David had and getting it up and working in the quickest way possible and then time willing working on improving the UX.

stephenwaite replied:
We just have to support multiple pt addr stat.

adunsulag replied:
Got all the style changes in to make this responsive.

adunsulag commented:
@bradymiller This is ready for a deeper review.  I will do some more extensive testing, but I've tested on fresh databases and existing databases and it appears to be working.  Its a lot of code but when you get a chance to review it, I'd appreciate it.  I'll fix the sql merge conflict tomorrow.

bradymiller commented on library/formatting.inc.php:
```diff
@@ -198,6 +173,7 @@ function DateFormatRead($mode = 'legacy')
 
 function DateToYYYYMMDD($DateValue)
 {
+    return DateFormatterUtils::DateToYYYYMMDD($DateValue);
```
shouldn't all the code below this line now be removed?

adunsulag replied:
Removed this code.

bradymiller commented on sql/6_1_0-to-7_0_0_upgrade.sql:
```diff
@@ -726,3 +726,64 @@ UPDATE `layout_options` SET `title` = 'Title' WHERE `layout_options`.`form_id` =
+SET @group_id = (SELECT `group_id` FROM layout_options WHERE field_id='street' AND form_id='DEM');
+SET @seq_add_to = (SELECT max(seq) FROM layout_options WHERE group_id = @group_id AND form_id='DEM');
+INSERT INTO `layout_options` (`form_id`, `field_id`, `group_id`, `title`, `seq`, `data_type`, `uor`, `fld_length`, `max_length`, `list_id`, `titlecols`, `datacols`, `default_value`, `edit_options`, `description`, `fld_rows`)
+VALUES ('DEM','additional_addresses',@group_id,'',@seq_add_to+1,54,1,0,0,'',4,4,'','','Additional Patient Addresses',0);
```
title is missing

adunsulag replied:
I did that intentionally to get the blank line effect that @psoas was going for.
![image](https://user-images.githubusercontent.com/228117/174305498-d4a6cd46-834f-4d10-8dd1-dca0ec1f115d.png)

bradymiller replied:
then sounds like also need to remove the title from database.sql entry

adunsulag replied:
Removed the title from database.sql

bradymiller commented on sql/6_1_0-to-7_0_0_upgrade.sql:
```diff
@@ -726,3 +726,64 @@ UPDATE `layout_options` SET `title` = 'Title' WHERE `layout_options`.`form_id` =
+    `priority` INT(11) NULL,
+    `type` VARCHAR(255) NULL COMMENT 'FK to list_options.option_id for list_id address-types',
+    `use` VARCHAR(255) NULL COMMENT 'FK to list_options.option_id for list_id address-uses',
+    `notes` TEXT(255) NULL,
```
text should be bare

bradymiller commented on sql/database.sql:
```diff
@@ -1137,6 +1138,41 @@ INSERT INTO `codes` (`code_text`,`code`,`code_type`) VALUES ('Inhale','C38216',1
+    `priority` INT(11) NULL,
+    `type` VARCHAR(255) NULL COMMENT 'FK to list_options.option_id for list_id address-types',
+    `use` VARCHAR(255) NULL COMMENT 'FK to list_options.option_id for list_id address-uses',
+    `notes` TEXT(255) NULL,
```
text should be bare

adunsulag replied:
k removed the null on the text column

bradymiller replied:
also need to remove the (255)

bradymiller commented:
@adunsulag , Looks great! Just noted a couple issues.

sjpadgett commented:
Pretty nice, Seems to test well for me. I'm adding the addresses array for generator address and about half done.

adunsulag commented:
@bradymiller @stephenwaite @sjpadgett Does anyone know what is up with the display mode column format of the LBF forms.  It doesn't seem to follow the same 4 column layout when displaying the data compared to when editing it.

Look for example here on the display
![image](https://user-images.githubusercontent.com/228117/174305774-c2cbd2a3-1c28-465c-b5cc-fb41a7738007.png)

Versus here on the edit
![image](https://user-images.githubusercontent.com/228117/174305881-798e27b8-1b76-4b4b-a61e-2eb260f45fd2.png)

The previous address is configured to have the label as 4 columns and the edit widget to have 4 columns.  The intent is to shove the previous addresses onto its own completely new line.

bradymiller commented on sql/6_1_0-to-7_0_0_upgrade.sql:
```diff
@@ -727,6 +727,67 @@ UPDATE `layout_options` SET `title` = 'Title' WHERE `layout_options`.`form_id` =
+    `priority` INT(11) NULL,
+    `type` VARCHAR(255) NULL COMMENT 'FK to list_options.option_id for list_id address-types',
+    `use` VARCHAR(255) NULL COMMENT 'FK to list_options.option_id for list_id address-uses',
+    `notes` TEXT(255),
```
also need to remove the (255)

mdsupport replied:
Is there any particular reason to use TEXT for `notes`? db engines tend to treat TEXT/blobs with storage related handling overheads. Is varchar's max size(21,844 utf8, 65,532 otherwise) not enough? This table should not have max rowsize issues.

sjpadgett replied:
should use tinytext.

adunsulag replied:
I don't particularly care one way or the other.  I will say that in most my designs I typically choose text columns once my column size gets over 256-512 characters.  MySQL would often inline a varchar into the row when joining column data if its under a certain size. Been a while since I've benchmarked this though.   (I just looked it up and they've bumped varchar inlining up to 768 bytes). 

I stuck with text as that was @psoas original design.  I don't know the data requirement on an address note, though 1024 characters is probably sufficient for this if we want it to stay as a varchar.  

Otherwise, I'm not sure what overhead you are referring to in treating text/blobs outside of the disk fetch when the column is included in the select.  But a varchar will have the same problem if it exceeds the inlining in the db engine.

adunsulag replied:
Changed this to be TinyText

sjpadgett commented:
```
<addr use="HP">
    <streetAddressLine>1357, Amber Dr</streetAddressLine>
    <city>Beaverton</city>
    <state>OR</state>
    <postalCode>97006</postalCode>
    <country>USA</country>
</addr>
<addr use="H">
    <streetAddressLine>100 Test Dr</streetAddressLine>
    <streetAddressLine>Room 2</streetAddressLine>
    <city>Orlando</city>
    <state>FL</state>
    <postalCode>33520</postalCode>
    <country>USA</country>
    <useablePeriod xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:type="IVL_TS">
        <low value="20170629"/>
        <high value="20220601"/>
    </useablePeriod>
</addr>
```

adunsulag commented:
@bradymiller so my last commit was dependent on this branch, so I just pushed it together as I wasn't sure if we were ready to bring this in yet or not.  I added in the fixes for the TinyText that @sjpadgett suggested.  

I also implemented the FHIR api support for multiple addresses.  I tested this against my local inferno on both a fresh db install and an existing db and that appears to be working.  Didn't test the bulk export but I figure if the standard API is working the bulk will work also.

Discovered a bug in the immunization api so I fixed that as part of this commit also.

bradymiller commented on sql/6_1_0-to-7_0_0_upgrade.sql:
```diff
@@ -922,3 +922,11 @@ UPDATE `list_options` SET `activity` = '1' WHERE `list_id` = 'race' AND `title`
 UPDATE `list_options` SET `activity` = '1' WHERE `list_id` = 'race' AND `title` = 'Asian Indian';
 #Endif
+
+#IfNotIndex patient_history idx_pid
```
fix (looks like it got inverted :) ):
```
#IfNotIndex patient_history pid_idx
```

adunsulag replied:
You know I just stared at this over and over again and couldn't figure out what you were saying, had to copy and paste the line and then my eyes saw the difference, lol.

bradymiller commented on sql/6_1_0-to-7_0_0_upgrade.sql:
```diff
@@ -922,3 +922,11 @@ UPDATE `list_options` SET `activity` = '1' WHERE `list_id` = 'race' AND `title`
+ALTER TABLE patient_history ADD INDEX `pid_idx` (`pid`);
+#EndIf
+
+#IfNotIndex contact_address
```
fix:
```
#IfNotIndex contact_address contact_address_idx
```

bradymiller commented on src/Services/PatientService.php:
```diff
@@ -16,7 +16,10 @@
 
 namespace OpenEMR\Services;
 
+use MongoDB\Driver\Query;
```
is this class being used here?

adunsulag replied:
Nope, removing.

bradymiller commented:
@adunsulag , Just noted 3 issues above. Also looks like it's failing the ci tests. When sorted out, then ready to bring into codebase :)

![2DV](https://user-images.githubusercontent.com/278968/174420922-eac6ddd9-ac30-4de3-b911-2fd4ef41d526.gif)

sjpadgett commented:
@adunsulag I thought I saw an issue with col for previous address rendering miscount between edit and view but can't find it!
Anyway a couple years ago I put in an edit option in layouts of Span Entire Row but didn't follow through.
I will now:
![image](https://user-images.githubusercontent.com/1859985/174448608-d036947b-0ed7-4947-8753-8aba95880c7f.png)
